<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Eoghan O’Halloran, Tiago Magalhães, Darren J. Fitzpatrick" />

<meta name="date" content="2016-09-24" />

<title>AncestryMapper Overview</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">AncestryMapper Overview</h1>
<h4 class="author"><em>Eoghan O’Halloran, Tiago Magalhães, Darren J. Fitzpatrick</em></h4>
<h4 class="date"><em>2016-09-24</em></h4>


<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#package-functions">Package Functions</a><ul>
<li><a href="#calculateamidsarith"><strong>calculateAMidsArith</strong></a></li>
<li><a href="#calculateamids"><strong>calculateAMids</strong></a><ul>
<li><a href="#producing-an-input-file-for-calculateamids">Producing an Input File for calculateAMids</a></li>
</ul></li>
<li><a href="#plotamids"><strong>plotAMids</strong></a></li>
<li><a href="#createmedoid"><strong>createMedoid</strong></a></li>
<li><a href="#refadd"><strong>refAdd</strong></a></li>
</ul></li>
<li><a href="#producing-a-sample-input-file">Producing a Sample Input File</a></li>
<li><a href="#example-data">Example Data</a></li>
<li><a href="#additional-data">Additional Data</a></li>
<li><a href="#tutorial">Tutorial</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>AncestryMapper is an R package that implements the methods described in “HGDP and HapMap Analysis by Ancestry Mapper Reveals Local and Global Population Relationships” Magalhães TR, Casey JP, Conroy J, Regan R, Fitzpatrick DJ, et al. PLoS ONE 7(11): e49438. (2012) Ancestry Mapper assigns genetic ancestry to an individual and allows the study of relationships between local and global populations. The method gives each individual an Ancestry Mapper Id (AMid), a genetic identifier comprising genetic coordinates that correspond to its relationship to various reference populations. The AMid metrics have intrinsic biological meaning and provide a tool to measure genetic similarity between world populations.</p>
</div>
<div id="package-functions" class="section level1">
<h1>Package Functions</h1>
<p>The package consists of four functions:</p>
<ul>
<li><p><strong>calculateAMidsArith</strong>: calculates and assigns Ancestry Mapper Ids (AMids) to each individual using a new, more precise, arithmetic method</p></li>
<li><p><strong>calculateAMids</strong>: calculates and assigns Ancestry Mapper Ids (AMids) identical to the older versions of AncestryMapper</p></li>
<li><p><strong>plotAMids</strong>: produces a heatmap representation of AMids</p></li>
<li><p><strong>createMedoid</strong>: constructs a reference from sample inputs</p></li>
<li><p><strong>refAdd</strong>: adds user-supplied references or data to a reference file, containing information for order, labelling and color</p></li>
</ul>
<div id="calculateamidsarith" class="section level2">
<h2><strong>calculateAMidsArith</strong></h2>
<p>For each individual, <strong>calculateAMidsArith</strong> computes the genetic distances amongst that individual and the set of selected references. As input, the function requires a tPED file, a standard file format used by the PLINK software suite. For details on the format see <a href="http://pngu.mgh.harvard.edu/~purcell/plink/#tped" class="uri">http://pngu.mgh.harvard.edu/~purcell/plink/#tped</a>.</p>
<p>It also requires the file ‘Corpheno’, containing the references with columns in the order of population, reference and order. An example of which can be found with the package.</p>
<table>
<thead>
<tr class="header">
<th align="left">Pheno_Pop</th>
<th align="left">Pheno_Data</th>
<th align="left">Pheno_Region</th>
<th align="left">Colors_Pop</th>
<th align="left">Colors_Region</th>
<th align="left">Colors_Data</th>
<th align="left">Order</th>
<th align="left">UNIQID</th>
<th align="left">Fam</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Example_Pop</td>
<td align="left">Fedorova2013</td>
<td align="left">EUR</td>
<td align="left">999999</td>
<td align="left">black</td>
<td align="left">black</td>
<td align="left">140</td>
<td align="left">Example_ID</td>
<td align="left">Example_Fam</td>
</tr>
<tr class="even">
<td align="left">Neanderthal.SP</td>
<td align="left">SP</td>
<td align="left">EUR</td>
<td align="left">999999</td>
<td align="left">black</td>
<td align="left">black</td>
<td align="left">140</td>
<td align="left">AltaiNea</td>
<td align="left">AltaiNea</td>
</tr>
<tr class="odd">
<td align="left">Denisovan.SP</td>
<td align="left">SP</td>
<td align="left">EUR</td>
<td align="left">999999</td>
<td align="left">black</td>
<td align="left">black</td>
<td align="left">140</td>
<td align="left">AltaiDen</td>
<td align="left">DenisovaPinky</td>
</tr>
</tbody>
</table>
<p>A file containing the record of the major and minor alleles as described by dbSNP is needed. An example file for use with the toy data can be found as ‘MinMaxFreq.rda’ included with the package.</p>
<p>Only SNPs with entries in the `MinMaxFreq.rda’ used will be used in the analysis. A larger MinMaxFreq file containing SNPs used on common Illumina and Affymetrix ChIPs in addition to others is available at: <a href="http://bit.ly/1OUstDP" class="uri">http://bit.ly/1OUstDP</a></p>
<p>Additionally the user can produce their own ‘MinMaxFreq’ reference file. This should be a data frame with 2 unnamed columns consisting of a column for each allele and row names corresponding to the rsID. This object must be named ‘MinMaxFreq’ and saved as an rda file.</p>
<table>
<thead>
<tr class="header">
<th align="left">row.names</th>
<th align="left">V1</th>
<th align="left">V2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">rs6663840</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="left">rs548726</td>
<td align="left">C</td>
<td align="left">T</td>
</tr>
<tr class="odd">
<td align="left">rs10803320</td>
<td align="left">C</td>
<td align="left">T</td>
</tr>
</tbody>
</table>
<p><strong>calculateAMidsArith</strong> returns a dataframe containing the genetic distance of each individual to the references. This provides the raw distance measures (Starting with the prefix C_) and indices (Normalized values, starting with the prefix I_).</p>
<p>The genetic distance is computed as the Euclidean distance normalized by the number of SNPs, between each individual and all the references used. AMids for a single individual from any dataset can be computed provided there is a reasonable overlap between the set of SNPs for that individual and all the references used. The AMids can take values from 0 to 2. In our experience, the values are in the range 0.4 to 1.1.</p>
<p>The normalized values of the distances are such that the highest reference is scored as 100, the lowest as 0 and all others adjusted accordingly. These indices place the individual in the genomic map, thus, they provide a global overview on the number of relevant references for each individual.</p>
</div>
<div id="calculateamids" class="section level2">
<h2><strong>calculateAMids</strong></h2>
<p>For each individual, calculateAMids computes the genetic distances amongst that individual and the set of HGDP references (or a set provided by the user). As input, the function requires a PED formatted file. PED formatting is the standard file format required by the PLINK software suite. For details on the format see <a href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#ped" class="uri">http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#ped</a>. It also requires a file containing the ids of the individuals to be used as references, and the population they correspond to.</p>
<p>As output, returns a dataframe containing the genetic distance of each individual to the all HGDP references. We provide the raw distance measures (starting with the prefix C_) and indices (normalized Values, starting with the prefix I_).</p>
<p>As with calculateAMidsArith, genetic distance is computed as the Euclidean distance normalized by the number of SNPs, between each individual and the references.</p>
<p>The user can include new references in AMids by editing the file ‘HGDP_References.txt’, inserting the population and the corresponding invidual’s name.</p>
<p>A demo file using 500 SNPs is included using data from the HGDP called ‘HGDP_500SNPs.ped’, this is compatiable with the ‘CorPheno’ file also included in the package. The small number of SNPs and samples will result in a highly noisy example.</p>
<p>For more information on using <strong>calculateAMids</strong> see, ?<strong>calculateAMids</strong>()</p>
<div id="producing-an-input-file-for-calculateamids" class="section level3">
<h3>Producing an Input File for calculateAMids</h3>
<p>The PED file should include individuals that will be taken as the population references, which will be used to calculate the ancestry mapper indexes (AMIds) for the user dataset. In our original work we used as references the 51 populations included in the Human Genome Diversity Project. The HGDP dataset can be obtained at <a href="http://hagsc.org/hgdp/files.html" class="uri">http://hagsc.org/hgdp/files.html</a>.</p>
<p>To merge a custom ped file with a ped file with the references, users could use PLINK. The commands --bmerge or --merge are used to merge two ped files. In most cases there will be strand inconsistencies, that can be rectified by flipping snps, using the command --flip. SNPs that are CG AT are impossible to determine which strand they are in and as such be removed. Ancestry mapper requires the ped files to be in the 1/2 coding system. The individual Ids are taken as the second column of the ped file; these ids should be unique.</p>
<p>We have produced a bed file with the references for the 51 HGDP populations, with 630,597 snps; the file is named HGDP_51RefAM_AutosomalSnps_630597_ACGT and can be obtained at <a href="http://bit.ly/2bkWDSQ" class="uri">http://bit.ly/2bkWDSQ</a>.</p>
</div>
</div>
<div id="plotamids" class="section level2">
<h2><strong>plotAMids</strong></h2>
<p><strong>plotAMids</strong> is used to visualize the relationship amongst individuals and the references. <strong>plotAMids</strong> takes as input the dataframe of genetic distances returned by <strong>calculateAMidsArith</strong>. The user can also provide a file with phenotypes for each individual which will be visible in the plot. Colors used were taken from the BlBrewer and RedBl packages.</p>
</div>
<div id="createmedoid" class="section level2">
<h2><strong>createMedoid</strong></h2>
<p><strong>createMedoid</strong> constructs an arithmetic reference from a tPED containing only one population with at least 10 individuals. The tPED should contain only individuals of a specific population and be formatted and prepared as described in ‘Producing a Sample Input File’.</p>
</div>
<div id="refadd" class="section level2">
<h2><strong>refAdd</strong></h2>
<p><strong>refAdd</strong> adds user-supplied references or data to the CorPheno file, giving population, color and ordering information to cluster the results of samples by population.</p>
</div>
</div>
<div id="producing-a-sample-input-file" class="section level1">
<h1>Producing a Sample Input File</h1>
<p>AncestryMapper’s references are produced using sequences in the same strand orientation as that used by dbSNP. Thus, all data input files for the functions <strong>calculateAMidsArith</strong> and <strong>createMedoid</strong> need also to be in the same orientation as dbSNP.</p>
<p>If your data is already in the same strand orientation as dbSNP you can skip this section.</p>
<p>The format for any sample data to be analysed by the function <strong>calculateAMidsArith</strong> or used to create a reference with <strong>createMedoid</strong> is a PLINK tPED.</p>
<p>All files used in this tutorial can be found in the ‘extdata’ folder with the package. An example BED file, ‘All-00’, containing the dbSNP alleles for 1000 SNPs is included.</p>
<p>A full version of the ‘All-00’ BED file covering 53,509,352 SNPs from the dbSNP database can be found at: <a href="http://bit.ly/1OUstDP" class="uri">http://bit.ly/1OUstDP</a></p>
<p>It is faster to use PLINK BED format for the following steps; at the end the BED file needs to be converted to tPED format.</p>
<p>It is faster to extract only those SNPs being used in your dataset from the All-00 BED and working with the output.</p>
<p>The most efficient way to get data in the same orientation as that used by dbSNP is to to merge the 00-All file with your BED file using the --merge command in PLINK; both files should be in the ACGT format.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">$plink --bfile All<span class="dv">-00</span>
 --bmerge Prep.bed Prep.bim Prep.fam
 --make-bed --out PrepMerge</code></pre></div>
<p>If there are SNPs in different strand orientations, a list of SNPs affected will be output in a file ending in ‘-merge.misssnp’.</p>
<p>Flip these SNPs using the file ending in ‘-merge.missnp’ in your sample file before merging the two files again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">$plink --bfile Prep --flip PrepMerge-merge.missnp --make-bed --out PrepFlipped

$plink --bfile All<span class="dv">-00</span>
 --bmerge PrepFlipped.bed PrepFlipped.bim PrepFlipped.fam
 --make-bed --out PrepMerge</code></pre></div>
<p>At the end of this merging, the ‘individual’ from the All-00 file should be removed using the --remove command in PLINK. A file containing the ID and family ID for the All-00 individual is included as `IndRm’</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">$plink --bfile PrepMerge --remove IndRm --make-bed --out PrepMerge</code></pre></div>
<p>SNPs that are CG or AT are invisible to the strand issue. The current AncestryMapper functions exclude them automatically from analysis.</p>
<p>Sites with any missingness are also automatically excluded from analysis. Thus, if you wish to use any SNPs with samples missing you will need to impute replacements or exclude individuals who have missing genotypes.</p>
<p>The final step is to transform the BED file to a tPED file and to output a list of the SNPs used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> $plink --bfile PrepMerge --transpose --recode --write-snplist --out PrepFinal</code></pre></div>
<p>It is important to keep a file with the SNPs used in the sample with the ending ‘.snplist’ as <strong>calculateAMidsArith</strong> will search for all the ‘.snplist’ files corresponding to the path of the tPEDs.</p>
</div>
<div id="example-data" class="section level1">
<h1>Example Data</h1>
<p>Example Data in the form of samples and references are provided containing 1000 SNPs with data from numerous populations and datasets containing 147 populations and 591 individuals. Due to the low number of SNPs, the plots will look noisy.</p>
</div>
<div id="additional-data" class="section level1">
<h1>Additional Data</h1>
<p>Full-sized medoids and a larger dbSNP reference (‘00-All’) for use with real user samples are currently being hosted at: <a href="http://bit.ly/1OUstDP" class="uri">http://bit.ly/1OUstDP</a></p>
</div>
<div id="tutorial" class="section level1">
<h1>Tutorial</h1>
<p>Below is a short tutorial</p>
<div id="section" class="section level4">
<h4></h4>
<p>The first step is to call the paths to the example data files distributed with the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(AncestryMapper)

<span class="co">#Path to folder containing population references</span>
Refs &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'data'</span>, <span class="dt">package =</span> <span class="st">'AncestryMapper'</span>)

<span class="co">#Path to folder containing samples in tPED format</span>
tpeds &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>, <span class="dt">package =</span> <span class="st">'AncestryMapper'</span>)

<span class="co">#Path to CorPheno file</span>
Corpheno &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>, <span class="st">'CorPheno'</span>, <span class="dt">package =</span> <span class="st">'AncestryMapper'</span>)

<span class="co">#Path to dbSNP allele data file</span>
All00Frq &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'data'</span>, <span class="st">'MinMaxFreq.rda'</span>, <span class="dt">package =</span> <span class="st">'AncestryMapper'</span>)</code></pre></div>
</div>
<div id="section-1" class="section level4">
<h4></h4>
<p>Calculate the genetic distance of the samples in the PLINK tPED files to the references.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genetic.distance &lt;-<span class="st"> </span><span class="kw">calculateAMidsArith</span>(<span class="dt">pathTotpeds =</span> tpeds,
       <span class="dt">NameOut =</span> <span class="st">'Example'</span>,
       <span class="dt">pathToAriMedoids =</span> Refs,
       <span class="dt">pathAll00 =</span> All00Frq)</code></pre></div>
</div>
<div id="section-2" class="section level4">
<h4></h4>
<p>The next step is to plot the results. In this example we are using the the normalised values for each individual. (columnPlot = ‘I’) For more information on the function arguments see ?plotAMids()</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotAMids</span>(<span class="dt">AMids =</span> genetic.distance, <span class="dt">phenoFile =</span> Corpheno, <span class="dt">columnPlot =</span> <span class="st">'I'</span>)</code></pre></div>
</div>
<div id="section-3" class="section level4">
<h4></h4>
<p>To plot samples by population or place them in a certain order, the samples need to be added to the CorPheno file.</p>
<p>If samples are not in the CorPheno file, the samples will be plotted under ‘Undefined’. If one or more references aren’t in the CorPheno, then they will be plotted at the start of the y axis before the entries with defined orders. An entry with any ID will suffice for references if the user does not wish to add or use the sample IDs from the data used to create the reference.</p>
<p>The CorPheno can easily be modified at the text level to change things such as population orders, colors or add new entries.</p>
<p>Alternatively the <strong>refAdd</strong> function may be more convenient. A tutorial on using the <strong>refAdd</strong>. For more information see ?refAdd()</p>
</div>
<div id="section-4" class="section level4">
<h4></h4>
<p>Users can also create their own population references from tPEDs, containing individuals from a specific population. This is done using the <strong>createMedoid</strong> function.</p>
<p>The required arguments are:</p>
<ul>
<li><p>pathTotpeds: The path to the tPED file(s), one tPED per population</p></li>
<li><p>pathAll00: The path to the file containing the dbSNP alleles</p></li>
</ul>
<p>Other options also exist and can be found with ?createMedoid()</p>
<p>Example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Path to folder containing samples for each population in tPED format</span>
tpeds &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;AncestryMapper&quot;</span>)

<span class="co">#Path to dbSNP allele data file</span>
All00Frq &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;MinMaxFreq.rda&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;AncestryMapper&quot;</span>)

<span class="kw">createMedoid</span>(<span class="dt">pathTotpeds =</span> tpeds, <span class="dt">pathAll00 =</span> All00Frq)</code></pre></div>
<p>The resulting rda file will be named with the prefix of ‘medoidArithmetic_’, followed by the name of the tPED used and the number of SNPs. E.g.</p>
<p>medoidArithmetic_Demo_1000_ChipMan.rda</p>
<p>In order for the new reference to be used, it will need to in the path specified.</p>
<p>For more information on the presence of ‘ChipMan’ at the end of the file name, see ?<strong>createMedoid</strong>()</p>
</div>
<div id="section-5" class="section level4">
<h4></h4>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
