---
title: "Genetic Recombination"
author: "Chris Gaynor"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Genetic Recombination}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This article provides an overview of the biology of genetic
recombination and how it is modeled within AlphaSimR. I will first
provide a brief review of meiosis with special emphasis on how process
differs between diploid and autopolyploid species. I will then discuss
genetic maps before concluding with an overview specific models used to
simulate genetic recombination.

## Meiosis

Usual process of diploids, two copies of each chromosome, one from
mother and one from father

Chromosome doubling, sister chromatids

Chromosome pairing in meiosis I, bivalent pair or tetrad

Chiasma form which may result in crossovers

Homologous pairs split

Sister chromatids split in meiosis II to produce haploid

### Meiosis in Polypoids

There are two primary types of polyploids: autopolyploids and
allopolyploids.

Allopolyploids, such as wheat, have multiple genome copies that come
from different species.

Meiosis and recombination proceeds similarly to diploids.

Autopolyploids have multiple copies of the same genome. This

## Genetic Maps

## Recombination Models

### Recombination Models in Autopolyploids

Introduction including basic details of inheritance

Review meiosis Stages of meiosis Chromosome pairing Chiasma (crossovers)
Crossover interference Segregation Diploids first and then polyploids
Centromeres for polyploids

Review genetic maps Development of maps in Morgan's fly lab Haldane
mapping function Kosambi mapping function

Recombination models Count-location model Gamma model Gamma sprinkling
model

Recombination tracking in AlphaSimR
