<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Coding-variant Allelic Series Test</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Coding-variant Allelic Series Test</h1>



<p>Updated: 2025-03-26</p>
<div id="data" class="section level2">
<h2>Data</h2>
<p>To run an allelic series test, there are 4 key inputs:</p>
<ul>
<li><p>A numeric annotation vector, of the same length as the number of
variants, coded as 1 for benign missense variants (BMVs), 2 for damaging
missense variants (DMVs), and 3 for protein truncating variants
(PTVs).</p></li>
<li><p>A covariates matrix, with as many rows as subjects and including
columns such as age and sex. If omitted, defaults to an intercept only.
If provided, an intercept will automatically be added if not
detected.</p></li>
<li><p>A genotype matrix, with subjects as rows and variants as columns.
The number of columns should correspond to the length of the annotation
vector.</p></li>
<li><p>A numeric phenotype vector, either continuous or binary.</p></li>
</ul>
<p>The example data used below were generated using the <code>DGP</code>
function provided with the package. The data set includes 100 subjects,
300 variants, and a continuous phenotype. The true effect sizes follow
an allelic series, with magnitudes proportional to
<code>c(1, 2, 3)</code> for BMVs, DMVs, and PTVs respectively.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">101</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="at">n =</span> n,</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="sc">/</span> <span class="fu">sqrt</span>(n),</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># Annotations.</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>anno <span class="ot">&lt;-</span> data<span class="sc">$</span>anno</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">head</span>(anno)</span></code></pre></div>
<pre><code>## rs1 rs2 rs3 rs4 rs5 rs6 
##   1   1   2   2   1   2</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Covariates.</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>covar <span class="ot">&lt;-</span> data<span class="sc">$</span>covar</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">head</span>(covar)</span></code></pre></div>
<pre><code>##      intercept         age sex          pc1        pc2         pc3
## [1,]         1 -0.77465210   1  0.539531581  1.2134116  0.63478280
## [2,]         1 -0.03492052   0  0.009270313  1.5256400 -0.62027221
## [3,]         1 -0.20782643   1 -0.777242512  1.2319750 -0.73311857
## [4,]         1  0.69939529   0 -0.357743663  0.1747664 -0.29754242
## [5,]         1  0.47420055   1 -1.083115262 -1.3005348 -0.68338808
## [6,]         1 -1.15038152   0 -0.539141451 -0.5159589  0.01971706</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Genotypes.</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>geno <span class="ot">&lt;-</span> data<span class="sc">$</span>geno</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">head</span>(geno[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>##      rs1 rs2 rs3 rs4 rs5
## [1,]   0   0   0   0   0
## [2,]   0   0   0   0   0
## [3,]   0   0   0   0   0
## [4,]   0   0   0   0   0
## [5,]   0   0   0   0   0
## [6,]   0   1   0   0   0</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Phenotype.</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> data<span class="sc">$</span>pheno</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">head</span>(pheno)</span></code></pre></div>
<pre><code>## [1] 0.2958143 1.7294107 0.4886318 1.0201730 1.0304183 1.9704552</code></pre>
<p>The example data generated by the preceding are available under
<code>vignettes/vignette_data</code>.</p>
</div>
<div id="running-the-alleic-series-test" class="section level2">
<h2>Running the alleic series test</h2>
<p>The COding-variant Allelic Series Test (COAST) is run using the
<code>COAST</code> function. By default, the output of
<code>COAST</code> includes a data.frame of counts showing the number of
alleles, variants, and carriers in each class that contributed to the
test, and a data.frame of p-values, with the <code>omni</code> test
denoting the final, overall p-value. Inspection of the component
p-values may be useful for determining which model(s) detected evidence
of an association. In the present case, the baseline count model
provided the greatest power.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>results <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">covar =</span> covar</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="fu">show</span>(results)</span></code></pre></div>
<pre><code>## Effect Sizes:
##         test beta    se
## 1       base 0.05 0.068
## 2       base 0.06 0.079
## 3       base 0.59 0.175
## 4        ind 0.31 0.256
## 5        ind 0.22 0.216
## 6        ind 0.66 0.218
## 7  max_count 0.06 0.048
## 8    max_ind 0.44 0.130
## 9  sum_count 0.07 0.031
## 10   sum_ind 0.19 0.058
## 
## 
## Counts:
##   anno alleles variants carriers
## 1    1     186      164       84
## 2    2     127      110       74
## 3    3      27       26       23
## 
## 
## P-values:
##           test   type     pval
## 1     baseline burden 6.62e-03
## 2          ind burden 9.11e-03
## 3    max_count burden 2.32e-01
## 4      max_ind burden 8.15e-04
## 5    sum_count burden 1.48e-02
## 6      sum_ind burden 1.09e-03
## 7 allelic_skat   skat 6.83e-02
## 8         omni   omni 4.68e-03</code></pre>
<p>The effect sizes data.frame is accessed via:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>results<span class="sc">@</span>Betas</span></code></pre></div>
<pre><code>##         test       beta         se
## 1       base 0.05309214 0.06836091
## 2       base 0.06427266 0.07917711
## 3       base 0.58612733 0.17472435
## 4        ind 0.30597919 0.25595872
## 5        ind 0.21943023 0.21607933
## 6        ind 0.66475646 0.21829831
## 7  max_count 0.05719340 0.04783470
## 8    max_ind 0.43655519 0.13041078
## 9  sum_count 0.07437087 0.03052869
## 10   sum_ind 0.19038906 0.05828998</code></pre>
<p>the counts data.frame via:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>results<span class="sc">@</span>Counts</span></code></pre></div>
<pre><code>##   anno alleles variants carriers
## 1    1     186      164       84
## 2    2     127      110       74
## 3    3      27       26       23</code></pre>
<p>and the p-values data.frame via:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>results<span class="sc">@</span>Pvals</span></code></pre></div>
<pre><code>##           test   type        pval
## 1     baseline burden 0.006618953
## 2          ind burden 0.009107121
## 3    max_count burden 0.231834498
## 4      max_ind burden 0.000815325
## 5    sum_count burden 0.014846685
## 6      sum_ind burden 0.001089858
## 7 allelic_skat   skat 0.068349669
## 8         omni   omni 0.004683243</code></pre>
<div id="ultra-rare-variants" class="section level3">
<h3>Ultra-rare variants</h3>
<p>Effect size estimation for ultra-rare variants is challenging due to
the scarcity of carriers on which to base the estimate. Following works
such as <a href="https://doi.org/10.1038/s41588-022-01178-w">SAIGE-GENE+</a>, we
recommend collapsing variants with a minor allele count (MAC) <span class="math inline">\(\leq 10\)</span> into an aggregate variant,
separately within each annotation category. The
<code>CollapseGeno</code> utility is provided for this purpose. The
<code>min_mac</code> specifies the threshold for retention as an
individual variant (e.g. <code>min_mac = 11</code> will collapse
variants with a MAC <span class="math inline">\(\leq 10\)</span>). The
output is a list containing the annotation vector <code>anno</code> and
genotype matrix <code>geno</code> after collapsing. In addition, a
mapping <code>vars</code> is provided showing which variants were
collapsed to create each aggregate variant. Note that:</p>
<ul>
<li><p><code>CollapseGeno</code> will change the order of the variants
in the genotype matrix.</p></li>
<li><p>No aggregate variant will be created within annotation categories
where no variants have a MAC below the threshold.</p></li>
<li><p>Collapsing ultra-rare variants does not guarantee that the
resulting aggregate variant will itself have a MAC <span class="math inline">\(\geq\)</span> the threshold. To ensure that only
variants with a minimum MAC enter the association test, a
<code>min_mac</code> threshold should also be supplied to
<code>COAST</code>.</p></li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">102</span>)</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co"># Generate data.</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fl">1e3</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>  <span class="at">n =</span> n,</span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">10</span>,</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">3</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>)</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co"># Collapse ultra-rare variants.</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>collapsed <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">CollapseGeno</span>(</span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>  <span class="at">anno =</span> data<span class="sc">$</span>anno, </span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>  <span class="at">geno =</span> data<span class="sc">$</span>geno,</span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>  <span class="at">min_mac =</span> <span class="dv">11</span></span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>)</span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a></span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a><span class="co"># Variants collapsed to form each aggregate variant.</span></span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a><span class="fu">head</span>(collapsed<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>##   anno            vars
## 1    1     rs1;rs4;rs5
## 2    2             rs8
## 3    3 rs2;rs3;rs7;rs9</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># Run COAST on the collapsed data.</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>results <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="at">anno =</span> collapsed<span class="sc">$</span>anno,</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  <span class="at">covar =</span> data<span class="sc">$</span>covar,</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>  <span class="at">geno =</span> collapsed<span class="sc">$</span>geno,</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>  <span class="at">pheno =</span> data<span class="sc">$</span>pheno,</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>  <span class="at">min_mac =</span> <span class="dv">10</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="different-numbers-of-annotation-categories" class="section level3">
<h3>Different numbers of annotation categories</h3>
<p><code>COAST</code> was originally intended to operate on the benign
missense variants, damaging missense variants, and protein truncating
variants within a gene, but it has been generalized to allow for an
arbitrary number of discrete annotation categories. The following
example simulates and analyzes data with 4 annotation categories. The
main difference when analyzing a different number of annotation
categories is that the <code>weight</code> vector should be specified,
and should have length equal to the number of <em>possible</em>
annotation categories. <code>COAST</code> will run, albeit with a
warning, if there are possible annotation categories to which no
variants are assigned (e.g. a gene contains no PTVs).</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">102</span>)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="co"># Generate data.</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fl">1e2</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>  <span class="at">n =</span> n,</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">400</span>,</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>) <span class="sc">/</span> <span class="fu">sqrt</span>(n),</span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>),</span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb20-11"><a href="#cb20-11" tabindex="-1"></a>)</span>
<span id="cb20-12"><a href="#cb20-12" tabindex="-1"></a></span>
<span id="cb20-13"><a href="#cb20-13" tabindex="-1"></a><span class="co"># Run COAST-SS.</span></span>
<span id="cb20-14"><a href="#cb20-14" tabindex="-1"></a>results <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb20-15"><a href="#cb20-15" tabindex="-1"></a>  <span class="at">anno =</span> data<span class="sc">$</span>anno,</span>
<span id="cb20-16"><a href="#cb20-16" tabindex="-1"></a>  <span class="at">covar =</span> data<span class="sc">$</span>covar,</span>
<span id="cb20-17"><a href="#cb20-17" tabindex="-1"></a>  <span class="at">geno =</span> data<span class="sc">$</span>geno,</span>
<span id="cb20-18"><a href="#cb20-18" tabindex="-1"></a>  <span class="at">pheno =</span> data<span class="sc">$</span>pheno,</span>
<span id="cb20-19"><a href="#cb20-19" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)</span>
<span id="cb20-20"><a href="#cb20-20" tabindex="-1"></a>)</span>
<span id="cb20-21"><a href="#cb20-21" tabindex="-1"></a><span class="fu">show</span>(results)</span></code></pre></div>
<pre><code>## Effect Sizes:
##         test  beta    se
## 1       base  0.01 0.065
## 2       base  0.23 0.062
## 3       base  0.19 0.096
## 4       base  0.61 0.116
## 5        ind -0.10 0.253
## 6        ind  0.53 0.205
## 7        ind  0.31 0.169
## 8        ind  0.96 0.186
## 9  max_count  0.18 0.035
## 10   max_ind  0.47 0.090
## 11 sum_count  0.11 0.018
## 12   sum_ind  0.19 0.035
## 
## 
## Counts:
##   anno alleles variants carriers
## 1    1     194      169       86
## 2    2     159      131       78
## 3    3      71       61       51
## 4    4      41       39       31
## 
## 
## P-values:
##           test   type     pval
## 1     baseline burden 9.67e-10
## 2          ind burden 5.86e-07
## 3    max_count burden 4.26e-07
## 4      max_ind burden 1.65e-07
## 5    sum_count burden 5.37e-10
## 6      sum_ind burden 8.24e-08
## 7 allelic_skat   skat 3.27e-06
## 8         omni   omni 4.11e-09</code></pre>
</div>
<div id="test-options" class="section level3">
<h3>Test options</h3>
<ul>
<li><code>apply_int = TRUE</code> applies the rank-based inverse normal
transformation from the <a href="https://CRAN.R-project.org/package=RNOmni">RNOmni</a> package.
This transformation is expected to improve power for phenotypes that
have a skewed or kurtotic (e.g. long-tailed) distribution. It is applied
by default in the case of continuous phenotype, and is ignored in the
case of a binary phenotype.</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>  <span class="at">apply_int =</span> <span class="cn">TRUE</span></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>include_orig_skato_all = TRUE</code> includes standard SKAT-O
applied to all variants as a component of the omnibus test, while
<code>include_orig_skato_ptv = TRUE</code> includes standard SKAT-O
applied to PTVs only. In cases where other annotation schemes are used,
the annotation for the PTV class can be specified via
<code>ptv_anno</code>. Including standard SKAT-O as a component of the
omnibus test can improve power to detect associations between the
phenotype and genes that may not be allelic series.</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>  <span class="at">include_orig_skato_all =</span> <span class="cn">TRUE</span>,</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>  <span class="at">include_orig_skato_ptv =</span> <span class="cn">TRUE</span>,</span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>  <span class="at">ptv_anno =</span> <span class="dv">3</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>is_pheno_binary = TRUE</code> is required to indicate that the
supplied phenotype is binary, and should be analyzed using a logistic
regression model.</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>  <span class="at">pheno =</span> <span class="dv">1</span> <span class="sc">*</span> (pheno <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>  <span class="at">is_pheno_binary =</span> <span class="cn">TRUE</span></span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>min_mac</code> is used to filter the variant set to those
containing a minimum minor allele count (MAC). The following example
filters to only those variants with a MAC of at least 2:</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>  <span class="at">min_mac =</span> <span class="dv">2</span></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>return_omni_only = TRUE</code> is used to return
<code>p_omni</code> only when the component p-values are not of
interest:</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>  <span class="at">return_omni_only =</span> <span class="cn">TRUE</span></span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>score_test = TRUE</code> specifies the use of a score-type
allelic series burden test. The default of
<code>score_test = FALSE</code> specifies a Wald-type allelic series
burden test.</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>  <span class="at">score_test =</span> <span class="cn">TRUE</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>)</span></code></pre></div>
<ul>
<li><code>weights</code> specifies the relative phenotypic effects of
BMVs, DMVs, and PTVs. An increasing pattern such as the default setting
of <code>weights = c(1, 2, 3)</code> targets allelic series. Setting
<code>weights = c(1, 1, 1)</code> would target a genetic architecture
where all variants have equivalent expected magnitudes.</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>  <span class="at">anno =</span> anno,</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>  <span class="at">geno =</span> geno,</span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>  <span class="at">pheno =</span> pheno,</span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>  <span class="at">covar =</span> covar,</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
