<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Covariate Inference</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Covariate Inference</h1>



<p>Updated: 2025-01-06</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(AllelicSeries)</span></code></pre></div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The primary goal of allelic series analysis is to determine whether a
dose-response relationship exists between the functionality of a gene
and the phenotype of interest. A secondary goal might be to quantify the
association between covariates and the phenotype in the context of an
allelic series model. This vignette will illustrate how to perform
inference on the latter relationship under two paradigms: the score test
setting and Wald test setting. The <strong>score test setting</strong>
estimates the association between the phenotype and covariates under the
null hypothesis of no genotypic effect. An advantage of score testing is
that all genes and all association models employed by COAST share a
common null model. Thus, the relationship between the phenotype and
covariates is characterized by a single set of association parameters.
Even if the null hypothesis does not hold for all genes, score testing
may be a reasonable approximation when the effect of genotype on the
phenotype is relatively small. By contrast, the <strong>Wald test
setting</strong> estimates the association between the phenotype and
covariates while allowing for a non-zero genotypic effect. Although more
flexible, a drawback of Wald testing is the need to estimate a separate
set of association parameters for each gene and each association model
employed by COAST.</p>
</div>
<div id="example-data" class="section level2">
<h2>Example data</h2>
<p>The simulated data provided by the <code>DGP</code> function includes
a covariate data matrix <code>covar</code> with columns representing an
intercept <code>int</code>, <code>age</code>, <code>sex</code>, and 3
genetic principal components <code>pc1</code>-<code>pc3</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">101</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(<span class="at">n =</span> <span class="fl">1e3</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">head</span>(data<span class="sc">$</span>covar)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt;      intercept          age sex        pc1        pc2         pc3</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; [1,]         1  0.003229867   1 -0.4170603  0.6779313  0.15930142</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; [2,]         1 -0.038306667   1 -1.1925043  0.7518853  0.83471897</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; [3,]         1 -0.399554232   1 -1.1574644 -0.2112389 -0.27238536</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; [4,]         1 -0.905447329   0 -0.2382042 -0.3219297  0.83435277</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; [5,]         1  0.637206113   1  1.0682585 -0.3522423  1.38841591</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; [6,]         1 -1.418399109   0  0.1814291  0.4061925  0.06444786</span></span></code></pre></div>
</div>
<div id="score-tests" class="section level2">
<h2>Score tests</h2>
<p>Score tests estimate the association between the phenotype and
covariates in the absence of a genotypic effect. The effect sizes can be
estimated by simple linear regression of the phenotype on covariates in
the case of a continuous phenotype, or logistic regression in the case
of a binary phenotype. Since all genes and all component allelic series
tests have a common null model, a single set of coefficients is
sufficient to characterize the association between the phenotype and
covariates.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Format score test data.frame.</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(data<span class="sc">$</span>covar)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>df<span class="sc">$</span>y <span class="ot">&lt;-</span> data<span class="sc">$</span>pheno</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># Case of a continuous phenotype.</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># An intercept is omitted from the call to `lm` because one is already </span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># contained in the covariate matrix.</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> ., <span class="at">data =</span> df)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt; lm(formula = y ~ 0 + ., data = df)</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">#&gt; Residuals:</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt;     Min      1Q  Median      3Q     Max </span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt; -3.2251 -1.1145 -0.3202  0.8159  7.1627 </span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt; intercept  0.95172    0.07279  13.075  &lt; 2e-16 ***</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co">#&gt; age       -0.07020    0.04993  -1.406   0.1600    </span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt; sex        0.14129    0.10026   1.409   0.1591    </span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="co">#&gt; pc1        0.04771    0.05007   0.953   0.3409    </span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co">#&gt; pc2        0.14178    0.04992   2.840   0.0046 ** </span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">#&gt; pc3        0.20038    0.04996   4.011  6.5e-05 ***</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a><span class="co">#&gt; ---</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a><span class="co">#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="co">#&gt; Residual standard error: 1.576 on 994 degrees of freedom</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a><span class="co">#&gt; Multiple R-squared:  0.3135, Adjusted R-squared:  0.3094 </span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a><span class="co">#&gt; F-statistic: 75.67 on 6 and 994 DF,  p-value: &lt; 2.2e-16</span></span></code></pre></div>
<p>The <strong>allelic series SKAT test</strong> is inherently a score
test. The <strong>allelic series burden tests</strong> may be run as
score tests by setting the <code>score_test</code> option to
<code>TRUE</code>, as in the following:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>results <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">COAST</span>(</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  <span class="at">anno =</span> data<span class="sc">$</span>anno,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">geno =</span> data<span class="sc">$</span>geno,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">pheno =</span> data<span class="sc">$</span>pheno,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">covar =</span> data<span class="sc">$</span>covar,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">score_test =</span> <span class="cn">TRUE</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="wald-tests" class="section level2">
<h2>Wald tests</h2>
<p>Wald tests estimate the association between the phenotype and
covariates allowing for the presence of a genotypic effect. We will
focus on estimation of effect sizes for the allelic series burden
models. The key difference from score testing is the need to calculate
the gene-burden score. This can be achieved using the
<code>Aggregator</code> function. The necessary inputs are the
annotation vector <code>anno</code> and genotype matrix
<code>geno</code>. The aggregation <code>method</code> argument should
be set to <code>&quot;none&quot;</code> for the baseline model (default),
<code>&quot;sum&quot;</code> for the allelic sum model, and <code>&quot;max&quot;</code> for
the allelic max model. The <code>indicator</code> argument should be set
to <code>FALSE</code> for additive genotype encoding (default), and
<code>TRUE</code> for dominance genotype encoding.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Example of fitting the baseline allelic series model.</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">Aggregator</span>(<span class="at">anno =</span> data<span class="sc">$</span>anno, <span class="at">geno =</span> data<span class="sc">$</span>geno, <span class="at">method =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">colnames</span>(g) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;g1&quot;</span>, <span class="st">&quot;g2&quot;</span>, <span class="st">&quot;g3&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>df_base <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">data.frame</span>(g), df)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> ., <span class="at">data =</span> df_base)</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; lm(formula = y ~ 0 + ., data = df_base)</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; Residuals:</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;      Min       1Q   Median       3Q      Max </span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; -2.25334 -0.53567  0.00053  0.49116  2.38950 </span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; g1         1.03675    0.05152  20.123  &lt; 2e-16 ***</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; g2         1.01910    0.02488  40.962  &lt; 2e-16 ***</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; g3         1.03262    0.03324  31.067  &lt; 2e-16 ***</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co">#&gt; intercept  0.06103    0.03912   1.560 0.119019    </span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co">#&gt; age       -0.09674    0.02438  -3.967 7.79e-05 ***</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co">#&gt; sex        0.09824    0.04915   1.999 0.045903 *  </span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="co">#&gt; pc1        0.09550    0.02447   3.903 0.000101 ***</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="co">#&gt; pc2        0.15654    0.02440   6.416 2.17e-10 ***</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a><span class="co">#&gt; pc3        0.26349    0.02447  10.768  &lt; 2e-16 ***</span></span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a><span class="co">#&gt; ---</span></span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a><span class="co">#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a><span class="co">#&gt; Residual standard error: 0.7694 on 991 degrees of freedom</span></span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a><span class="co">#&gt; Multiple R-squared:  0.837,  Adjusted R-squared:  0.8355 </span></span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a><span class="co">#&gt; F-statistic: 565.4 on 9 and 991 DF,  p-value: &lt; 2.2e-16</span></span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a><span class="co"># Example of fitting the allelic series sum model.</span></span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">Aggregator</span>(<span class="at">anno =</span> data<span class="sc">$</span>anno, <span class="at">geno =</span> data<span class="sc">$</span>geno, <span class="at">method =</span> <span class="st">&quot;sum&quot;</span>)</span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">colnames</span>(g) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;g_sum&quot;</span>)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a>df_sum <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">data.frame</span>(g), df)</span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> ., <span class="at">data =</span> df_sum)</span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a><span class="co">#&gt; lm(formula = y ~ 0 + ., data = df_sum)</span></span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a><span class="co">#&gt; Residuals:</span></span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="co">#&gt;      Min       1Q   Median       3Q      Max </span></span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a><span class="co">#&gt; -2.25447 -0.53172  0.00076  0.49099  2.38847 </span></span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a><span class="co">#&gt; g_sum      1.02571    0.01817  56.463  &lt; 2e-16 ***</span></span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a><span class="co">#&gt; intercept  0.06204    0.03883   1.598 0.110432    </span></span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a><span class="co">#&gt; age       -0.09659    0.02435  -3.967 7.80e-05 ***</span></span>
<span id="cb5-52"><a href="#cb5-52" tabindex="-1"></a><span class="co">#&gt; sex        0.09744    0.04889   1.993 0.046540 *  </span></span>
<span id="cb5-53"><a href="#cb5-53" tabindex="-1"></a><span class="co">#&gt; pc1        0.09534    0.02443   3.903 0.000102 ***</span></span>
<span id="cb5-54"><a href="#cb5-54" tabindex="-1"></a><span class="co">#&gt; pc2        0.15644    0.02434   6.427 2.02e-10 ***</span></span>
<span id="cb5-55"><a href="#cb5-55" tabindex="-1"></a><span class="co">#&gt; pc3        0.26281    0.02438  10.778  &lt; 2e-16 ***</span></span>
<span id="cb5-56"><a href="#cb5-56" tabindex="-1"></a><span class="co">#&gt; ---</span></span>
<span id="cb5-57"><a href="#cb5-57" tabindex="-1"></a><span class="co">#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb5-58"><a href="#cb5-58" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-59"><a href="#cb5-59" tabindex="-1"></a><span class="co">#&gt; Residual standard error: 0.7687 on 993 degrees of freedom</span></span>
<span id="cb5-60"><a href="#cb5-60" tabindex="-1"></a><span class="co">#&gt; Multiple R-squared:  0.837,  Adjusted R-squared:  0.8358 </span></span>
<span id="cb5-61"><a href="#cb5-61" tabindex="-1"></a><span class="co">#&gt; F-statistic: 728.3 on 7 and 993 DF,  p-value: &lt; 2.2e-16</span></span>
<span id="cb5-62"><a href="#cb5-62" tabindex="-1"></a></span>
<span id="cb5-63"><a href="#cb5-63" tabindex="-1"></a><span class="co"># Example of fitting the allelic series max model.</span></span>
<span id="cb5-64"><a href="#cb5-64" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">Aggregator</span>(<span class="at">anno =</span> data<span class="sc">$</span>anno, <span class="at">geno =</span> data<span class="sc">$</span>geno, <span class="at">method =</span> <span class="st">&quot;max&quot;</span>)</span>
<span id="cb5-65"><a href="#cb5-65" tabindex="-1"></a><span class="fu">colnames</span>(g) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;g_max&quot;</span>)</span>
<span id="cb5-66"><a href="#cb5-66" tabindex="-1"></a>df_max <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">data.frame</span>(g), df)</span>
<span id="cb5-67"><a href="#cb5-67" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> ., <span class="at">data =</span> df_max)</span>
<span id="cb5-68"><a href="#cb5-68" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb5-69"><a href="#cb5-69" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-70"><a href="#cb5-70" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb5-71"><a href="#cb5-71" tabindex="-1"></a><span class="co">#&gt; lm(formula = y ~ 0 + ., data = df_max)</span></span>
<span id="cb5-72"><a href="#cb5-72" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-73"><a href="#cb5-73" tabindex="-1"></a><span class="co">#&gt; Residuals:</span></span>
<span id="cb5-74"><a href="#cb5-74" tabindex="-1"></a><span class="co">#&gt;     Min      1Q  Median      3Q     Max </span></span>
<span id="cb5-75"><a href="#cb5-75" tabindex="-1"></a><span class="co">#&gt; -2.2191 -0.5904 -0.0163  0.5239  4.3631 </span></span>
<span id="cb5-76"><a href="#cb5-76" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-77"><a href="#cb5-77" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb5-78"><a href="#cb5-78" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb5-79"><a href="#cb5-79" tabindex="-1"></a><span class="co">#&gt; g_max      1.16910    0.02421  48.283  &lt; 2e-16 ***</span></span>
<span id="cb5-80"><a href="#cb5-80" tabindex="-1"></a><span class="co">#&gt; intercept  0.05036    0.04396   1.145 0.252332    </span></span>
<span id="cb5-81"><a href="#cb5-81" tabindex="-1"></a><span class="co">#&gt; age       -0.08737    0.02730  -3.200 0.001418 ** </span></span>
<span id="cb5-82"><a href="#cb5-82" tabindex="-1"></a><span class="co">#&gt; sex        0.09045    0.05484   1.649 0.099378 .  </span></span>
<span id="cb5-83"><a href="#cb5-83" tabindex="-1"></a><span class="co">#&gt; pc1        0.10294    0.02741   3.756 0.000183 ***</span></span>
<span id="cb5-84"><a href="#cb5-84" tabindex="-1"></a><span class="co">#&gt; pc2        0.15302    0.02730   5.605 2.69e-08 ***</span></span>
<span id="cb5-85"><a href="#cb5-85" tabindex="-1"></a><span class="co">#&gt; pc3        0.23371    0.02733   8.552  &lt; 2e-16 ***</span></span>
<span id="cb5-86"><a href="#cb5-86" tabindex="-1"></a><span class="co">#&gt; ---</span></span>
<span id="cb5-87"><a href="#cb5-87" tabindex="-1"></a><span class="co">#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb5-88"><a href="#cb5-88" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-89"><a href="#cb5-89" tabindex="-1"></a><span class="co">#&gt; Residual standard error: 0.8621 on 993 degrees of freedom</span></span>
<span id="cb5-90"><a href="#cb5-90" tabindex="-1"></a><span class="co">#&gt; Multiple R-squared:  0.7949, Adjusted R-squared:  0.7935 </span></span>
<span id="cb5-91"><a href="#cb5-91" tabindex="-1"></a><span class="co">#&gt; F-statistic: 549.9 on 7 and 993 DF,  p-value: &lt; 2.2e-16</span></span></code></pre></div>
<p>Effect sizes for the allelic series SKAT model may be estimated by
fitting a (generalized) linear mixed-effects model with an appropriately
specified random-effect for genotype. However, fitting such models is
uncommon, as the SKAT test is inherently a score test, and developing a
strategy for Wald-type estimation of the fixed-effects in a SKAT model
is beyond the scope of this vignette.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
