<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Data Generation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Data Generation</h1>



<p>Updated: 2024-11-12</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(AllelicSeries)</span></code></pre></div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The data generating process <code>DGP</code> provides the ability to
simulate data from a variety of genetic architectures, including the
baseline model, allelic sum models, and allelic SKAT models.</p>
<div id="baseline-model" class="section level3">
<h3>Baseline model</h3>
<p>The <em>baseline allelic series model</em> is: <span class="math display">\[ \mathbb{E}(Y|N,X) = \sum_{l=1}^{L}N_{l}\beta_{l}
+ X&#39;\beta_{X} \]</span> where <span class="math inline">\(Y\)</span>
is the phenotype, <span class="math inline">\(L\)</span> is the number
of annotation categories, <span class="math inline">\(N_{l}\)</span> is
the number of variants in annotation category <span class="math inline">\(l\)</span>, and <span class="math inline">\(X\)</span> is a vector of covariates with
associated coefficient <span class="math inline">\(\beta_{X}\)</span>.
To simulate from the baseline model, the aggregation <code>method</code>
is set to <code>&quot;none&quot;</code>. <code>n</code> is the sample size, and
<code>snps</code> in the number of variants. <code>prop_anno</code>
specifies the proportion of variants in each annotation category. <span class="math inline">\(L = 3\)</span> annotation categories are adopted
by default. <code>beta</code> is the per-variant effect size in each
annotation category. <code>weights</code> is redundant with
<code>beta</code> in the case of the baseline model, but these arguments
have distinct functions in the allelic sum and max models.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.1</span>),</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="allelic-sum-model" class="section level3">
<h3>Allelic sum model</h3>
<p>The <em>allelic series sum model</em> is: <span class="math display">\[ \mathbb{E}(Y|N,X) =
\left(\sum_{l=1}^{L}N_{l}w_{l}\right)\beta + X&#39;\beta_{X} \]</span>
To simulate from the sum model, the aggregation <code>method</code> is
set to <code>&quot;sum&quot;</code>. In contrast to the baseline model,
<code>beta</code> is now a scalar multiplier for the allelic sum burden
<span class="math inline">\(\sum_{l=1}^{L}N_{l}w_{l}\)</span>, while
<code>weights</code> specifies the annotation category weights <span class="math inline">\((w_{1}, \dots, w_{L})\)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;sum&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.1</span>),</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">beta =</span> <span class="dv">1</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="allelic-max-model" class="section level3">
<h3>Allelic max model</h3>
<p>The <em>allelic series max model</em> is: <span class="math display">\[ \mathbb{E}(Y|N,X) =
\left(\max_{l=1}^{L}N_{l}w_{l}\right)\beta + X&#39;\beta_{X}
\]</span></p>
<p>To simulate from the max model, the aggregation <code>method</code>
is set to <code>&quot;max&quot;</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;max&quot;</span>,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.1</span>),</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">beta =</span> <span class="dv">1</span>,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="allelic-skat-model" class="section level3">
<h3>Allelic SKAT model</h3>
<p>The generative version of the <em>allelic series SKAT model</em> is:
<span class="math display">\[ \begin{gathered}
\mathbb{E}(Y|G,X) = \sum_{j=1}^{J}G_{j}\beta_{j} + X&#39;\beta_{X}, \\
\beta_{j} = r_{j}\gamma_{j}\beta_{A_{j}}
\end{gathered} \]</span></p>
<p>Here <span class="math inline">\(G_{j}\)</span> is genotype at the
<span class="math inline">\(J\)</span>th rare variant and <span class="math inline">\(\beta_{j}\)</span> is the corresponding effect
size. The effect size of each variant are the product of a random sign
<span class="math inline">\(r_{j} \in \{-1, 1\}\)</span>, a scalar
frailty <span class="math inline">\(\gamma_{j} \sim \Gamma(\alpha,
\alpha)\)</span> with mean 1 and variance <span class="math inline">\(\alpha^{-1}\)</span>, and <span class="math inline">\(\beta_{A_{j}}\)</span> is the mean absolute effect
size for variant <span class="math inline">\(j\)</span>’s annotation
category <span class="math inline">\(A_{j} \in \{1, \dots, L\}\)</span>.
To simulate from the SKAT model, the aggregation <code>method</code> is
set to <code>&quot;none&quot;</code> and the <code>random_signs</code> argument is
set to <code>TRUE</code>. The mean absolute effect sizes are set via
<code>beta</code>. The variance of the frailty <span class="math inline">\(\gamma_{j}\)</span> is specified with
<code>random_var</code>. While the annotation category
<code>weights</code> are not explicitly required for generation from the
SKAT model, <code>weights</code> should be provided because the number
of annotation categories <span class="math inline">\(L\)</span> is
inferred from the length of the weight vector.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">random_signs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">random_var =</span> <span class="dv">1</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.1</span>),</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="simulating-more-than-3-annotation-categories" class="section level3">
<h3>Simulating more than 3 annotation categories</h3>
<p>Data can be simulated from any of the baseline, sum, max, or SKAT
models simply by changing the lengths of annotation category
proportions, effect sizes, and weights. These arguments should all be
updated to reflect the number of annotation categories <span class="math inline">\(L\)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Baseline model, 2 categories.</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.6</span>, <span class="fl">0.4</span>),</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>),</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>)</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co"># Baseline model, 4 categories.</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>data <span class="ot">&lt;-</span> AllelicSeries<span class="sc">::</span><span class="fu">DGP</span>(</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">100</span>,</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  <span class="at">snps =</span> <span class="dv">300</span>,</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>  <span class="at">prop_anno =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>),</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>  <span class="at">weights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="additional-arguments" class="section level3">
<h3>Additional arguments</h3>
<ul>
<li><p>Real-data annotations or genotypes can be provided by specifying
<code>anno</code> or <code>geno</code>.</p></li>
<li><p>To simulate binary phenotypes from a probit model, set
<code>binary = TRUE</code>. The binary phenotype is generated by first
constructing a latent normal phenotype <code>Z</code> then dichotomizing
<span class="math inline">\(Y = \mathbb{I}(Z &gt; 0)\)</span>.</p></li>
<li><p>The range of minor allele frequencies can be specified with
<code>maf_range</code>. By default, variants have MAFs in the range of
0.1% to 0.5%. Genotypes are generated in such a way that the minor
allele count is always at least 1, guaranteeing that no empty variants
will be present.</p></li>
<li><p>The proportion of causal variants can be modulated with
<code>prop_causal</code>. By default, all variants are causal for the
phenotype (unless <code>beta</code> is set to zero). If
<code>prop_causal &lt; 1</code>, then a corresponding proportion of
variants is removed from the causal set.</p></li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
