<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>introduction</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">introduction</h1>



<div id="load-your-single-cell-objects" class="section level4">
<h4><strong>Load your single cell object(s)</strong></h4>
<p>AnanseSeurat requires a preprocessed single cell object. Before
exporting data make sure the single cell object has been filtered for
good quality cells, and has underwent sufficient pre-processing. For
guides on pre-processing scRNA-seq and scATAC-seq there are some great
vignettes and guides available. We will also set the working
directory.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># if (!requireNamespace(&quot;remotes&quot;, quietly = TRUE)) {</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#   install.packages(&quot;remotes&quot;)</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># Sys.unsetenv(&quot;GITHUB_PAT&quot;)</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># remotes::install_github(&quot;JGASmits/AnanseSeurat&quot;)</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(AnanseSeurat)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(Signac)</span></code></pre></div>
<p>In this example we will use a preprocsed 10x PBMC multiome dataset
(PBMC from a Healthy Donor (v1, 150x150) Single Cell Multiome ATAC +
Gene Expression Dataset by Cell Ranger ARC 2.0.0). This data was
pre-processed following the standard pre-processing from Signac. The
pre-pocessed single cell object is available at Zenado: <a href="https://zenodo.org/record/7446267/" class="uri">https://zenodo.org/record/7446267/</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>rds_file <span class="ot">&lt;-</span> <span class="st">&#39;preprocessed_PDMC.Rds&#39;</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(rds_file)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p>This single cell object contains multimodal data, both RNA and ATAC
signal from each cell. However in the case of a separate scRNA-seq
object and scATAC-seq object, AnanseSeurat can still prepare files for
running single cell Ananse. In the case of two seperate objects as input
it is important that both the objects share their respective cluster
names.</p>
</div>
<div id="output-files-using-ananseseurat" class="section level4">
<h4><strong>Output files using AnanseSeurat</strong></h4>
<p>We will start with outputting the CPM file, for this we select:</p>
<ol style="list-style-type: decimal">
<li>the minimum amount of cells a cluster needs to have to be included
via min_cells</li>
<li>the output directory</li>
<li>the cluster ID containing the cluster names</li>
<li>the name of the Assay containing the RNA data.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">export_CPM_scANANSE</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  pbmc,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  min_cells <span class="ot">&lt;-</span> <span class="dv">25</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">output_dir =</span> <span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&#39;/analysis&#39;</span>),</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">RNA_count_assay =</span> <span class="st">&#39;RNA&#39;</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>)</span></code></pre></div>
<p>Next we will output the ATAC peak matrix, , for this we select:</p>
<ol style="list-style-type: decimal">
<li>the minimum amount of cells a cluster needs to have to be included
via min_cells</li>
<li>the output directory</li>
<li>the cluster ID containing the cluster names</li>
<li>the name of the Assay containing the ATAC peak data.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">export_ATAC_scANANSE</span>(</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  pbmc,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  min_cells <span class="ot">&lt;-</span> <span class="dv">25</span>,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">output_dir =</span> <span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&#39;/analysis&#39;</span>),</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">ATAC_peak_assay =</span> <span class="st">&#39;peaks&#39;</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>)</span></code></pre></div>
<p>Next we will generate the config and sample file needed for
anansnake, we will also specify specific pairwise comparisons between
clusters of interest. By default, scANANSE compares all clusters to a
network based on the average values of all clusters. Additional
comparisons can be specified, in this case, B-naive and B-memory cells
were also specified to compare directly to each other.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>contrasts <span class="ot">&lt;-</span>  <span class="fu">list</span>(<span class="st">&#39;B-naive_B-memory&#39;</span>,</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>                   <span class="st">&#39;B-memory_B-naive&#39;</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                   <span class="st">&#39;B-naive_CD16-Mono&#39;</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                   <span class="st">&#39;CD16-Mono_B-naive&#39;</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">config_scANANSE</span>(</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  pbmc,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  min_cells <span class="ot">&lt;-</span> <span class="dv">25</span>,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>  <span class="at">output_dir =</span> <span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&#39;/analysis&#39;</span>),</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>  <span class="at">genome =</span> <span class="st">&#39;./data/hg38&#39;</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>  <span class="at">additional_contrasts =</span> contrasts</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>)</span></code></pre></div>
<p>Finally we will calculate the markergenes for each cluster and
between clusters of a specific comparison:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">DEGS_scANANSE</span>(</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>  pbmc,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  min_cells <span class="ot">&lt;-</span> <span class="dv">25</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">output_dir =</span> <span class="st">&#39;./analysis&#39;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">additional_contrasts =</span> contrasts</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="run-anansnake-on-the-generated-files" class="section level4">
<h4><strong>run Anansnake on the generated files</strong></h4>
<p>After this all your files are ready to run Anansnake for gene
regulatory network analysis. For info on installing and running
anansnake see: <a href="https://github.com/vanheeringen-lab/anansnake" class="uri">https://github.com/vanheeringen-lab/anansnake</a></p>
</div>
<div id="import-influence-scores-back-into-your-single-cell-object" class="section level4">
<h4><strong>Import influence scores back into your single cell
object</strong></h4>
<p>After running anansnake the results can be incorporated back into
your single cell object.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">import_seurat_scANANSE</span>(pbmc,</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>                               <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                               <span class="at">anansnake_inf_dir =</span> <span class="st">&quot;./analysis/influence/&quot;</span>)</span></code></pre></div>
<p>The top TFs contributing to your clusters can be vizualized as a
table:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>TF_influence <span class="ot">&lt;-</span> <span class="fu">per_cluster_df</span>(pbmc,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                               <span class="at">assay =</span> <span class="st">&#39;influence&#39;</span>,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                               <span class="at">cluster_id =</span> <span class="st">&#39;predicted.id&#39;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">head</span>(TF_influence)</span></code></pre></div>
<p>We can also visualize the influence score of TFs in the single cell
object:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>highlight_TF1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;STAT4&#39;</span>, <span class="st">&#39;MEF2C&#39;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">DefaultAssay</span>(<span class="at">object =</span> pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>plot_expression1 <span class="ot">&lt;-</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(pbmc, <span class="at">features =</span> highlight_TF1, <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="fu">DefaultAssay</span>(<span class="at">object =</span> pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;influence&quot;</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>plot_ANANSE1 <span class="ot">&lt;-</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>    pbmc,</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">1</span>,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>    <span class="at">features =</span> highlight_TF1,</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;darkgrey&quot;</span>, <span class="st">&quot;#fc8d59&quot;</span>)</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  )</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="fu">print</span>(plot_expression1 <span class="sc">|</span> plot_ANANSE1)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
